### ISSUE TITLE
Pytorch support

### ISSUE THREAD

**Issue by JonathanSum** [pyodide/pyodide#1625]
Does pyodide support Pytorch?
Or does wasm support Pytorch?

**Comment by rth** [comment]
If you want to do inference the easiest would likely be to convert the model to ONNX then use it in WASM with ONNX.js

See the upstream issue: https://github.com/pytorch/pytorch/issues/25091

Building Pytorch in Pyodide is theoretically possible however looking at the [list of required build and run dependencies in conda-forge](https://github.com/conda-forge/pytorch-cpu-feedstock/blob/master/recipe/meta.yaml) it would be very complex and time consuming. Also it would be CPU only.

So projects like ONNX.js would be more appropriate because there you would also be able to use the GPU via WebGL.

Happy to keep this issue open for future reference, but it's unlikely we would be able to do much about it on the Pyodide side.

**Comment by JonathanSum** [comment]
ONNX doesn't work. ONNX javascript does not support all the functions of PyTorch, and, most importantly, it does not support some custom functions using in the PyTorch model. I have already built a project with ONNX, and I feel it is not really a good solution.

I hope pyodide will have a certain way to support Pytorch, so I can deploy something like a custom transformer model. Or it may allow me to train it on GPU.

I will close this in the next 2 days, but I still think it should at least support Pytorch for machine learning people.

**Comment by rth** [comment]
Yes, I forgot I previously answered this question in  https://stackoverflow.com/a/64407516/1791279 and there is indeed a blocker on ctypes / CFFI support (https://github.com/pyodide/pyodide/issues/728)

> ONNX javascript does not support all the functions of PyTorch, and, most importantly, it does not support some custom functions using in the PyTorch model.

ONNX.js is indeed outdated, however there is apparently a WASM backend for onnxruntime directly https://github.com/microsoft/onnxjs/issues/292#issuecomment-840091143

Have you tried raising that issue with them? For classical Language Models with pytorch it works quite well.  

I understand the interest, it's just that doing this is a very large project and we have fairly limited resources.

**Comment by fs-eire** [comment]
> ONNX doesn't work. ONNX javascript does not support all the functions of PyTorch, and, most importantly, it does not support some custom functions using in the PyTorch model. And it looks like a dead project at its repo(https://github.com/microsoft/onnxjs)
> 
> I have already built a project with ONNX, and I feel it is not really a good solution.
> 
> I hope pyodide will have a certain way to support Pytorch, so I can deploy something like a custom transformer model. Or it may allow me to train it on GPU.
> 
> In addition, this post has said that Pyodide does not support Pytorch because of Pyodide itself.
> https://stackoverflow.com/questions/64358372/run-pytorch-in-pyodide
> 
> Again, I believe if it supports Pytorch, it will shrine-like Reactjs because Pytorch is easier to use. And the deploying issue is what Pytorch lacks.

ONNX Runtime Web, as a successor of ONNX.js. is release in ONNX Runtime v1.8.0 (2021-06-02). This uses WebAssembly to compile the whole ONNX Runtime and supports all operators in latest ONNX opset. please check it at https://github.com/microsoft/onnxruntime/tree/master/js/web.

**Comment by daoxian** [comment]
> ONNX doesn't work. ONNX javascript does not support all the functions of PyTorch, and, most importantly, it does not support some custom functions using in the PyTorch model. And it looks like a dead project at its repo(https://github.com/microsoft/onnxjs)
> 
> I have already built a project with ONNX, and I feel it is not really a good solution.
> 
> I hope pyodide will have a certain way to support Pytorch, so I can deploy something like a custom transformer model. Or it may allow me to train it on GPU.
> 
> In addition, this post has said that Pyodide does not support Pytorch because of Pyodide itself.
> https://stackoverflow.com/questions/64358372/run-pytorch-in-pyodide
> 
> Again, I believe if it supports Pytorch, it will shrine-like Reactjs because Pytorch is easier to use. And the deploying issue is what Pytorch lacks.

@JonathanSum You could try tensorflow.js or Alibaba MNN.js for inference, which can make use of GPU.  Pyodide is suitable for pre and post processing instead of inference.

**Comment by lsb** [comment]
Question: if the blockers for pytorch had been cffi support and ctypes support, and both are resolved, are there any other blockers to running pytorch via pyodide? Is this something reasonable to attempt?

**Comment by ryanking13** [comment]
> Question: if the blockers for pytorch had been cffi support and ctypes support, and both are resolved ( https://github.com/pyodide/pyodide/tree/main/packages/cffi and #1656), are there any other blockers to running pytorch via pyodide? Is this something reasonable to attempt?

@lsb I once tried to build PyTorch via pyodide naively with most options disabled, but it was unsuccessful at that time.
PyTorch is big and messy; there are a bunch of third-party libraries, torch and caffe2 codes are mixed up, and the combination of some build options is unreliable. I still have interests in building PyTorch, but I'm not working on it right now. So there can be unknown blockers.

**Comment by msaroufim** [comment]
> > Question: if the blockers for pytorch had been cffi support and ctypes support, and both are resolved ( https://github.com/pyodide/pyodide/tree/main/packages/cffi and #1656), are there any other blockers to running pytorch via pyodide? Is this something reasonable to attempt?
> 
> @lsb I once tried to build PyTorch via pyodide naively with most options disabled, but it was unsuccessful at that time. PyTorch is big and messy; there are a bunch of third-party libraries, torch and caffe2 codes are mixed up, and the combination of some build options is unreliable. I still have interests in building PyTorch, but I'm not working on it right now. So there can be unknown blockers.

Flagging this to @seemethere @malfet

**Comment by donno2048** [comment]
:eyes:

**Comment by tchaton** [comment]
Any progress ?

**Comment by ryanking13** [comment]
@tchaton Sadly no. There are no known blockers, but it's all a matter of time and effort.

Anyone who is familiar with the build system of Torch and interested in supporting this can give it a try, then we can help debugging the issues. As you seem like a member of Pytorch Lightning team, you might be the best person for the job if you are interested.

**Comment by ValMystletainn** [comment]
I have tried to build the pytorch wheel by pyodide build.  I have overcome some difficulties by changing the config and patch to source file. But I am stuck on the support of pthread.
According to #237 . there are no pthread support right now in pyodide? If I understand it correctly.
However, it's seems not easy to aovid the useage of pthread in pytorch (supporting protobuf, some threading accelerated operator).

I'm posting this comment to check if I'm in the right place to do this work

**Comment by ryanking13** [comment]
@ValMystletainn

Thanks for sharing. Yes, pthread is not supported in Pyodide, and I don't think we will be able to support it in the near future. So I guess you'll need to contact PyTorch team to find a way to disable pthread support.

**Comment by Tvenus** [comment]
Hi
i am python developer.
Could you please give me an opportunity to work with you?

**Comment by ryanking13** [comment]
@Tvenus Sure, feel free to work on this issue and share us your results. I don't think there is a person who is actively working on this issue, so sadly there isn't much thing we can do for you, but we will be happy to help you debug the issues you are facing.
